utils/warnings.ml

```diff
@@ -383,7 +526,19 @@ let parse_opt error active errflag s =
        loop (i+1)
     | _ -> error ()
   in
-  loop 0
+  if s = "" then ()
+  else begin
+    match name_to_number (String.sub s 1 (String.length s - 1)), s.[0] with
+    | None, _ ->
+        begin match name_to_number s with
+        | None -> loop 0
+        | Some n -> set n
+        end
+    | Some n, '+' -> set n
+    | Some n, '-' -> clear n
+    | Some n, '@' -> set_all n
+    | _ -> loop 0
+  end
```

Comment:
I believe that this is subtly wrong if we have warnings named `foo` and also `oo`. You should check `s` before `rest`.

(Nitpick: I would match on `s.[0], name_to_number rest` rather than in the other order, so that source order follows input order.)

Reply:
Good catch! Fixed.